### PR TITLE
[Handshake] Add pass that locks functions

### PR DESCRIPTION
**Pull request by Dinistro**
This PR introduces a handshake pass that adds a lock mechanism for functions.
Such a mechanism ensures that there is only one active control token in a function at each point in time.

An additional "RUN" directive/command was added to the integration tests that otherwise require a task-pipelining transformation. These test runs disable task-pipelining in favor of locking but should still yield the same results (albeit with lower throughput).

### CODE REVIEW

**Comment by teqdruid** [comment]
I'm curious: what is this useful for (other than debugging)?

**Comment by Dinistro** [comment]
Handshake functions produced by `std-to-handshake` (with task-pipelining disabled) do not guarantee correct execution when multiple control tokens are live at the same time. Instead of performing a transformation that involves loop detections and other graph traversals, this pass simply constructs a lock around the region.

**Comment by mikeurbach** [comment]
At a high level, I think this makes sense. Historically, we wrote the driver code (ie. integration tests, or whatever hooked up the top-level handshake module) to only assert the control input for one cycle, and wait until the control output was asserted before launching another invocation. Then the task pipelining work came along. But I think this protection makes sense for a simple guard in the absence of task pipelining.

**Comment by teqdruid** [comment]
Kinda seems like it aught to be called all the time when task pipelining is turned off. (But I know nothing about the handshake passes.)

File: lib/Dialect/Handshake/Transforms/LockFunctions.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  // Dummy value that causes a buffer initialization, but itself does not have a
+  // semantic meaning.
+  buff->setAttr("initValues", rewriter.getI64ArrayAttr({0}));
+
+  auto ctrl = entry->getArguments().back();
+  auto join = rewriter.create<JoinOp>(loc, ValueRange({ctrl, buff}));
+  ctrl.replaceAllUsesExcept(join, join);
+
+  auto ret = *r.getOps<handshake::ReturnOp>().begin();
```
Comment (mortbopet) [inline lib/Dialect/Handshake/Transforms/LockFunctions.cpp:45]: As an alternative, `getTerminator(...)` should be applicable here.

**Review by mortbopet — APPROVED** [review]
LGTM!

**Comment by Dinistro** [comment]
> Kinda seems like it aught to be called all the time when task pipelining is turned off. (But I know nothing about the handshake passes.)

If one wants to enforce that all the circuits produced by `std-to-handshake` are safe, then we can do that. I'll change it, as there is currently no obvious use case for a separate pass. Should one arise, later on, changing this back to a separated pass isn't much effort.

**Comment by mortbopet** [comment]
>  as there is currently no obvious use case for a separate pass.

I would disagree here - we already have passes which are mandatory for correct execution, but separate (fork/sink materialization, bufferization) - and it is then expected that the "user" (in reality, a tool which composes a DHLS flow) to run the correct passes in the correct order. I'm generally an advocate for this approach, especially when it comes to `standardToHandshake` since it's already somewhat of a monolithic pass.

**Comment by Dinistro** [comment]
I do understand both reasons. The locking pass can be applied to all handshake functions and is independent of the remainder of `std-to-handshake`. On the other hand, it is mainly used for circuits that are produced with task pipelining disabled (pipelining vs. locking). 

The main benefit of having a separate pass is the possibility to test it in isolation. The `std-to-handshake` tests are hard to reason about. Splitting large transformations into multiple steps allows localizing bugs a lot simpler.

**Comment by Dinistro** [comment]
To address the issue of managing a large set of "micro-passes", Morten and I suggest a `hlstool`: https://github.com/llvm/circt/issues/3864

**Comment by Dinistro** [comment]
Are there any objections regarding moving this into a separate pass?

**Pull request merged**